### PR TITLE
Fix 3 bugs in gpfdist and file protocol in fstream.c

### PR DESCRIPTION
**Pull request by zhaorui20005**
First, this pr adds auto detecting eol if header option is specified.
Second, both file and gpfdist protocols are using “fstream_read” to read data from local files. But the legacy codes did not check if the file ends with the correct eol. So when there are multiple files in the fstream to handle, the last line content in the previous file will concatenate with the first line in the next file.
The code paths of file protocol and gpfdist protocol are not the same. So this pr fixes 3 bugs.
There are many corner cases if we check if the file ends with the correct eol and try to add them to the outputs.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/misc/fstream/fstream.c [hunk 1]
```diff
@@ -699,6 +704,143 @@ char* format_error(char* c1, char* c2)
 	return err_msg;
 }
 
+#define set_line_delim(eol_type) \
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/fstream.c:707]: why use macro instead of inline function?

File: src/backend/utils/misc/fstream/fstream.c [hunk 2]
```diff
@@ -699,6 +704,143 @@ char* format_error(char* c1, char* c2)
...
+		default:
+			buff[bytesread] = '\n';
+			ret = 1;
+			break;
+	}
+	return ret;
+}
+static char *find_nc_in_text_buffer(char *p, char *q, int nc)
+{
+	char*	last_record_loc = 0;
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/fstream.c:787]: NULL is better than 0 for pointer.

File: src/backend/utils/misc/fstream/fstream.c [hunk 3]
```diff
@@ -699,6 +704,143 @@ char* format_error(char* c1, char* c2)
...
+	while (p < q)
+	{
+		int ch = *p++;
+
+		if (ch == nc)
+		{
+			last_record_loc = p;
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/fstream.c:794]: so last_record_loc is the location of char behind nc?

File: src/backend/utils/misc/fstream/fstream.c (hunk 1 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:707]: Because I don't want to write the passing parameters. It needs at least 3 parameters.

File: src/backend/utils/misc/fstream/fstream.c (hunk 2 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:787]: Sure. I will modify this line.

File: src/backend/utils/misc/fstream/fstream.c (hunk 3 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:794]: Yes, we only need the relative position of cr and lf. So we don't need to plus one.

File: src/backend/utils/misc/fstream/fstream.c [hunk 4]
```diff
@@ -699,6 +704,143 @@ char* format_error(char* c1, char* c2)
...
+			break;
+		case EOL_NL:
+		default:
+			buff[bytesread] = '\n';
+			ret = 1;
+			break;
+	}
+	return ret;
+}
+static char *find_nc_in_text_buffer(char *p, char *q, int nc)
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/fstream.c:785]: does it work if there is \r or \n in quotation?
I think we needn't parse data to find newline if the eol_type is unknown.

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:785]: When \r or \n in quotation, it is not regarded as a real eol, so the the find_line_delimiter_known will find the first \r or \n according to the format. So if it is a 'csv' format, it will find the first \r or \n outside the quotation in function scan_csv_records_cr_or_lf.
At first I didn't find the eol for eol_type unknown. But there is a case in gpload failed. The eol is \r but without \n in the file. So the code append a \n at the end which caused it failed.

Comment (huiliang-liu) [reply src/backend/utils/misc/fstream/fstream.c:785]: but what about text format? is it also correct?
I think it is better that split this PR as two or three PRs. one PR for one bug. It will be clear to be reviewed and we can merge them one by one.

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:785]: Yes, the "find_line_delimiter_known" function will seperate the 'csv' and 'text' format in if and else path.
Actually the first bug can be easily solved by setting the eol_string and eol_length for file protocol in header path. 
And the second and the third bugs are the same problems, but in different protocols(file and gpfdist protocols).

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:785]: Both protocols are using the same fstream_read function. The gpfdist will automatically switch to nextfile when the reading bytes are less than expected. But in file protocol, only when the gfile_read  return 0, it will switch to the next file. So the fixing will be slightly different.
For gpfdist, we only need to care about the last read buffer. So we don't need to store last char. But for file protocol, only when gfile_read returns 0 that we will try to add eol, but the last buffers are return from last read, so we need to store last char.

File: src/backend/utils/misc/fstream/fstream.c (hunk 1 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/fstream.c:707]: Also it need to pass a pointer to a char pointer.

**Comment by haolinw** [comment]
@zhaorui20005 Is this PR still actionable ? I saw it's been stay for months.

**Comment by zhaorui20005** [comment]
> @zhaorui20005 Is this PR still actionable ? I saw it's been stay for months.

This pr really fix something, but a bit complicated, and there is no need from customers. So I think we could close this for now, and reopen other prs if customer complains about the related problems.
